### DISCUSSION TITLE
Allowed Domains Setting doesn't seem to Work

### DISCUSSION THREAD

**Discussion by Dan6erbond** [outline/outline#4083] [Self Hosting]
I have Outline hosted in a Kubernetes cluster, and configured with Authentik OIDC for SSO. So far this has worked well with @gmail.com Users, until I noticed that some colleagues who wanted to access with @gmx.ch, etc. were out of luck. Based on my research I then tried to add those domains to the Allowed Domains setting in Outline, but logging in with those users still doesn't work. Is this expected behavior as Outline's SSO might not cover custom OIDC providers or something else?

**Comment by tommoor** [comment]
- What version?
- What specifically does "doesn't work" mean?

**Reply by Dan6erbond** [reply]
I'm using version 0.66.0 and the behavior in general seems to be a bit weird around authentication. Granted; I accidentally created two users with the same email, so Outline initially would always change their names when they logged-in, but gave them the same account. Once I fixed that, getting users into Outline with an @gmx.ch email address would show this error in the UI:

> Authentication failed – you do not have permission to access this team.

And in the logs I see this:

```
{"error":"Invalid authentication","stack":"UnauthorizedError: Invalid authentication\n    at InvalidAuthenticationError (/opt/outline/build/server/errors.js:48:34)\n    at userProvisioner (/opt/outline/build/server/commands/userProvisioner.js:156:50)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:85:20)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:87:22)","level":"error","message":"Error during authentication"}
```

Sometimes I also get this error, which I am having just as much trouble explaining as the issue with domains:

```
{"error":"Authentication required","stack":"UnauthorizedError: Authentication required\n    at AuthenticationError (/opt/outline/build/server/errors.js:41:34)\n    at OIDCClient.rotateToken (/opt/outline/build/server/utils/oauth.js:75:47)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async Object.refreshAccessTokenIfNeeded (/opt/outline/build/server/models/UserAuthentication.js:157:24)\n    at async Object.validateAccess (/opt/outline/build/server/models/UserAuthentication.js:110:7)\n    at async Promise.all (index 0)\n    at async /opt/outline/build/server/queues/tasks/ValidateSSOAccessTask.js:40:21\n    at async /opt/outline/node_modules/sequelize/lib/sequelize.js:503:18\n    at async ValidateSSOAccessTask.perform (/opt/outline/build/server/queues/tasks/ValidateSSOAccessTask.js:26:5)\n    at async Queue.<anonymous> (/opt/outline/build/server/services/worker.js:132:7)","level":"error","message":"Error processing task in ValidateSSOAccessTask"}
```

And this one:

```
{"error":"The provided authorization grant or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client","stack":"TokenError: The provided authorization grant or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client\n    at OAuth2Strategy.parseErrorResponse (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:373:12)\n    at OAuth2Strategy._createOAuthError (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:420:16)\n    at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:177:45\n    at /opt/outline/node_modules/oauth/lib/oauth2.js:191:18\n    at passBackControl (/opt/outline/node_modules/oauth/lib/oauth2.js:132:9)\n    at IncomingMessage.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:157:7)\n    at IncomingMessage.emit (node:events:538:35)\n    at IncomingMessage.emit (node:domain:475:12)\n    at endReadableNT (node:internal/streams/readable:1345:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","level":"error","message":"Error during authentication"}
```

For reference, my test user looks like this:

![image](https://user-images.githubusercontent.com/32235074/189059838-5e77439d-23e2-4ead-90a5-4bccfd263e16.png)

And the Authentik provider configuration:

![image](https://user-images.githubusercontent.com/32235074/189059989-1b914381-4c2a-434a-86a7-d3e49a725e3d.png)

![image](https://user-images.githubusercontent.com/32235074/189060040-b78c2b7a-0203-41f7-b884-55ad2e4ec9c1.png)

Here are the security settings in Outline:

![image](https://user-images.githubusercontent.com/32235074/189060090-4b13e946-b19e-4c7a-b00e-da4161fe19f2.png)

**Reply by tommoor** [reply]
My guess is that you have `DEPLOYMENT=hosted` set in your environment somehow which would drastically alter the login behavior.

I'd also recommend increasing the validity of tokens to days or weeks rather than minutes.

**Reply by Dan6erbond** [reply]
What is the purpose of the `DEPLOYMENT` environment variable in general, and what are its possible values? I can't recall explicitly setting it, so that might just be left at its default value. But I will check. Thank you for the answer!

**Reply by tommoor** [reply]
It's for getoutline.com and should never be set to any value in community edition

**Reply by Dan6erbond** [reply]
I checked the deployment. The `DEPLOYMENT` environment variable doesn't seem to be set, and additionally I just found that I cannot even add `outlook.com` as an allowed domain, which is really limiting behavior as currently I don't need Outline to decide who can and cannot access my wiki, since that's Authentik's job.

But when I remove the Allowed Domains entries from the settings altogether, it only seems to allow gmail.com addresses... My guess is that it defaults to the one that the first user/admin has?

**Reply by Dan6erbond** [reply]
Looking at the most recent release, and comparing to a previous version I had deployed locally I think I might have gotten a bit closer to the source of this error.

v0.66.0 adds #3813 which seems to use the email matching to join teams, but in my case this might result in Outline believing that a team doesn't exist for all emails other than gmail.com (the initial - thus default) and not allowing the user to join any team at all.

Can it be confirmed that this change is causing this behavior? It would definitely explain the headaches, and will allow me to consider whether I should downgrade to v0.65.0 or if a quick fix is planned - assuming this isn't intended behavior for selfhosted Outline.

**Reply by tommoor** [reply]
Allowed domains is totally optional, if your SSO method already restricts access then you can remove all values in allowed domains and there should be no restrictions in place other than being able to sign in with SSO. 

Recent changes should only have made auth _more_ flexible, not less.

**Reply by Dan6erbond** [reply]
I see. I also tried removing the Allowed domains, which was initially the case anyway, and still got errors, which is why I tried setting that in case it was anything related to the configuration.

But as I mentioned, either way I get an error. If Allowed domains is configured, users that have a different domain get this error:

> Sorry, your domain is not allowed. Please try again with an allowed team domain.

If they do have a domain found in the Allowed domains setting, or allowed domains isn't set, they get this:

> Authentication failed - you do not have permission to access this team.

So my guess is that #3813 kind of broke the login flow, because currently the workaround that I found, that works, but isn't very convenient, is first logging them into my team by going to the Team Settings > Members > Invite people and adding their email.

**Reply by heilerich** [reply]
We have the exact same problem with our installation since upgrading to v0.66.0. As @Dan6erbond mentioned the only workaround right now is to invite everyone manually. I think we should move this to an issue. This is definitely a bug.

**Reply by tommoor** [reply]
cc @thenanyu

**Reply by alex-drakos** [reply]
@tommoor we're seeing the same exact problem on a hosted / xxxx.getoutline.com instance: even though allowed domains are setup, trying to SSO with an email that was not the same domain originally used to create the Team results in a `invalid-authentication` error.

**Reply by tommoor** [reply]
It is possible that this setting is enabled…

<img width="537" alt="image" src="https://user-images.githubusercontent.com/380914/191111797-26894654-7b4e-4322-b2df-84ecbf351d20.png">

**Reply by alex-drakos** [reply]
No, that's not the case with us (and also tried toggling on/off a few other of the security options). Just FYI, we do have two main email domains and each is on its own Google Workspace account, but I'd excpect by setting the allowed domains that it would work with SSO of that domain.
We tried inviting people with the secondary domain email and that worked, but would be great if SSO sign ups worked as with the primary email domain.

**Reply by tommoor** [reply]
Ah, some of the work in this PR should enable that soon: https://github.com/outline/outline/pull/3964 – we're removing the hard SSO <> Team relationship

**Reply by lenaxia** [reply]
> It is possible that this setting is enabled…
> 
> <img alt="image" width="537" src="https://user-images.githubusercontent.com/380914/191111797-26894654-7b4e-4322-b2df-84ecbf351d20.png">

Is this option available in the self-hosted version? I'm on 0.66.1 and I don't see this option under security.

**Reply by tommoor** [reply]
The option does not exist in self hosted

**Reply by reddec** [reply]
@tommoor is it possible to get list of all differences between self hosted and cloud version? I was sure that project is fully open-source

**Reply by tommoor** [reply]
The “self managed” tab on pricing page mostly reflects the differences:

https://www.getoutline.com/pricing
